pythonlib/src/mill/pythonlib/PythonModule.scala

```diff
@@ -21,59 +40,81 @@ trait PythonModule extends Module {
     PathRef(python)
   }
 
-  def typeCheck: T[Unit] = Task {
-    Task.traverse(moduleDeps)(_.typeCheck)()
-
-    os.call(
-      (pythonExe().path, "-m", "mypy", "--strict", sources().path),
-      stdout = os.Inherit,
-      cwd = T.workspace
+  // TODO: right now, any task that calls this helper will have its own python
+  // cache. This is slow. Look into sharing the cache between tasks.
+  protected def runPythonExe(args: Task[Seq[String]]): Task[Unit] = Task.Anon {
+    Jvm.runSubprocess(
+      commandArgs = Seq(pythonExe().path.toString) ++ args(),
+      envArgs = Map(
+        "PYTHONPATH" -> transitiveSources().map(_.path).mkString(java.io.File.pathSeparator),
+        "PYTHONPYCACHEPREFIX" -> (T.dest / "cache").toString,
+        if (Task.log.colored) { "FORCE_COLOR" -> "1" }
+        else { "NO_COLOR" -> "1" }
+      ),
+      workingDir = Task.workspace
     )
   }
 
-  def gatherScripts(upstream: Seq[(PathRef, PythonModule)]) = {
-    for ((sourcesFolder, mod) <- upstream) {
-      val destinationPath =
-        os.pwd / mod.millSourcePath.subRelativeTo(mill.api.WorkspaceRoot.workspaceRoot)
-      os.copy.over(sourcesFolder.path / os.up, destinationPath)
-    }
+  def typeCheck: T[Unit] = Task {
+    Task.traverse(moduleDeps)(_.typeCheck)()
```

Comment:
Looking at this again, since we do not actually use the upstream `typeCheck` results, should we skip this line? That would allow all typechecking to happen in parallel even for dependent modules, which would be nice

Reply:
indeed. We'll also need to look into generating stub files so that typechecking works when importing other modules defined in the same build. Right now `mypy` errors out when refering to other modules, lest we explicitly ignore import definitions. I think however that this can be done in a second PR